File: packages/sdk-ios/Sources/TraceItX/SharedData/SharedData.swift
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2026 ScriptX
//
// Loads packages/protocol/data/*.json (mirrored into Sources/TraceItX/Resources/
// at scaffold time and CI-gated for drift) into the SDK at first-access. Eager
// `static let` evaluation means each file is parsed exactly once per process
// lifetime, then cached. No locks needed — Swift `let`-static initialization is
// thread-safe at the language level.
//
// Per CONTEXT decision 4: shared data JSON is the cross-language source of
// truth. Each native SDK reimplements logic idiomatically but consumes the
// same JSON, so iOS / Android / web cannot drift on what counts as "sensitive".
import Foundation

// Resource bundle resolver. The published artifact is the xcframework
// (project.yml → scripts/build-xcframework.sh), where Bundle(for:) keyed
// off a private marker class resolves to the framework's own bundle. The
// SPM target is kept ONLY for the `swift test` dev workflow; there
// `Bundle.module` is the synthetic SPM accessor that points at the
// resource sub-bundle SPM builds alongside the module.
#if SWIFT_PACKAGE
private let traceitxResourceBundle: Bundle = .module
#else
private final class TraceItXResourceMarker {}
private let traceitxResourceBundle: Bundle = Bundle(for: TraceItXResourceMarker.self)
#endif

public struct RedactionPattern: Codable, Equatable, Sendable {
    public let id: String
    public let regex: String
    public let replacement: String?
    public let description: String?
}

public struct SensitiveHeaders: Codable, Sendable {
    public let version: Int
    public let sensitive_to_redact: [String]
    public let allowed_to_capture: [String]
}

public struct NetworkPathPatterns: Codable, Sendable {
    public let version: Int
    public let sensitive_paths: [SensitivePath]
    public struct SensitivePath: Codable, Sendable {
        public let id: String
        public let regex: String
        public let redact: String
        public let replacement: String?
    }
}

public enum SharedData {
    public static let redactionPatterns: [RedactionPattern] = {
        guard let url = traceitxResourceBundle.url(forResource: "redaction-patterns", withExtension: "json"),
              let data = try? Data(contentsOf: url) else { return [] }
        struct Wrapper: Codable { let patterns: [RedactionPattern] }
        return (try? JSONDecoder().decode(Wrapper.self, from: data).patterns) ?? []
    }()

    public static let sensitiveHeadersToRedact: Set<String> = {
        guard let url = traceitxResourceBundle.url(forResource: "sensitive-headers", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let decoded = try? JSONDecoder().decode(SensitiveHeaders.self, from: data) else { return [] }
        return Set(decoded.sensitive_to_redact.map { $0.lowercased() })
    }()

    public static let allowedHeadersToCapture: Set<String> = {
        guard let url = traceitxResourceBundle.url(forResource: "sensitive-headers", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let decoded = try? JSONDecoder().decode(SensitiveHeaders.self, from: data) else { return [] }
        return Set(decoded.allowed_to_capture.map { $0.lowercased() })
    }()

    public static let networkPathPatterns: NetworkPathPatterns? = {
        guard let url = traceitxResourceBundle.url(forResource: "network-path-patterns", withExtension: "json"),
              let data = try? Data(contentsOf: url) else { return nil }
        return try? JSONDecoder().decode(NetworkPathPatterns.self, from: data)
    }()
}

File: packages/sdk-android/android/everframe-core/src/main/kotlin/dev/everframe/shared/SharedData.kt
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2026 ScriptX
//
// Loads packages/protocol/data/*.json (mirrored into packages/sdk-android/android/everframe-core/src/main/assets/everframe/
// by the `copyProtocolData` Gradle task) into the SDK at first-access. Lazy-init
// means each file is parsed exactly once per process, then cached.
//
// Per CONTEXT decision 4: shared data JSON is the cross-language source of
// truth. iOS / Android / web cannot drift on what counts as "sensitive".
//
// Mirrors `packages/sdk-ios/Sources/Everframe/SharedData/SharedData.swift`.
package dev.everframe.shared

import android.content.Context
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class RedactionPattern(
    val id: String,
    val regex: String,
    val replacement: String? = null,
    val description: String? = null,
)

@Serializable
internal data class RedactionPatternsFile(
    val version: Int = 1,
    val description: String? = null,
    val patterns: List<RedactionPattern>,
)

@Serializable
data class SensitiveHeadersData(
    val version: Int = 1,
    val description: String? = null,
    @SerialName("sensitive_to_redact") val sensitiveToRedact: List<String>,
    @SerialName("allowed_to_capture") val allowedToCapture: List<String>,
)

@Serializable
data class NetworkPathPattern(
    val id: String,
    val regex: String,
    val redact: String,
    val replacement: String? = null,
)

@Serializable
internal data class NetworkPathPatternsFile(
    val version: Int = 1,
    val description: String? = null,
    @SerialName("sensitive_paths") val sensitivePaths: List<NetworkPathPattern>,
)

object SharedData {
    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = false
    }

    @Volatile
    private var appContext: Context? = null

    /**
     * Initialize with the host's application context. Plan 05-02 wires this
     * inside `Everframe.start(context, config)` before captureGate flips.
     */
    fun init(context: Context) {
        appContext = context.applicationContext
    }

    /**
     * Test/diagnostic seam — reset cached parsed JSON so a subsequent call
     * re-reads from assets. Production code does not call this.
     */
    @androidx.annotation.VisibleForTesting
    internal fun __resetForTesting() {
        appContext = null
        _redactionPatterns = null
        _sensitiveHeaders = null
        _networkPathPatterns = null
        _sensitiveToRedactLower = null
        _allowedToCaptureLower = null
    }

    @Volatile private var _redactionPatterns: List<RedactionPattern>? = null
    @Volatile private var _sensitiveHeaders: SensitiveHeadersData? = null
    @Volatile private var _networkPathPatterns: List<NetworkPathPattern>? = null
    @Volatile private var _sensitiveToRedactLower: Set<String>? = null
    @Volatile private var _allowedToCaptureLower: Set<String>? = null

    val redactionPatterns: List<RedactionPattern>
        get() = _redactionPatterns ?: synchronized(this) {
            _redactionPatterns ?: loadJson("redaction-patterns.json") { txt ->
                json.decodeFromString<RedactionPatternsFile>(txt).patterns
            }.also { _redactionPatterns = it }
        }

    val sensitiveHeaders: SensitiveHeadersData
        get() = _sensitiveHeaders ?: synchronized(this) {
            _sensitiveHeaders ?: loadJson("sensitive-headers.json") { txt ->
                json.decodeFromString<SensitiveHeadersData>(txt)
            }.also { _sensitiveHeaders = it }
        }

    val networkPathPatterns: List<NetworkPathPattern>
        get() = _networkPathPatterns ?: synchronized(this) {
            _networkPathPatterns ?: loadJson("network-path-patterns.json") { txt ->
                json.decodeFromString<NetworkPathPatternsFile>(txt).sensitivePaths
            }.also { _networkPathPatterns = it }
        }

    /** Lower-cased view of the sensitive header names — case-insensitive lookup. */
    val sensitiveHeadersToRedact: Set<String>
        get() = _sensitiveToRedactLower ?: synchronized(this) {
            _sensitiveToRedactLower
                ?: sensitiveHeaders.sensitiveToRedact.map { it.lowercase() }.toSet()
                    .also { _sensitiveToRedactLower = it }
        }

    /** Lower-cased view of the allowlist header names — case-insensitive lookup. */
    val allowedHeadersToCapture: Set<String>
        get() = _allowedToCaptureLower ?: synchronized(this) {
            _allowedToCaptureLower
                ?: sensitiveHeaders.allowedToCapture.map { it.lowercase() }.toSet()
                    .also { _allowedToCaptureLower = it }
        }

    private inline fun <T> loadJson(name: String, parse: (String) -> T): T {
        val ctx = appContext ?: error(
            "SharedData not initialized — call SharedData.init(context) before reading shared JSON. " +
                "Everframe.start(context, config) handles this automatically."
        )
        val text = ctx.assets.open("everframe/$name").bufferedReader().use { it.readText() }
        return parse(text)
    }
}
